pkg/services/anonymous/anonimpl/anonstore/database.go

```diff
@@ -0,0 +1,148 @@
+		args = append(args, from.UTC(), to.UTC())
+	}
+	err := s.sqlStore.GetSqlxSession().Select(ctx, &devices, query, args...)
+	if err != nil {
+		return nil, err
+	}
+	return devices, nil
+}
+
+func (s *AnonDBStore) CreateOrUpdateDevice(ctx context.Context, device *Device) error {
+	var query string
+
+	args := []any{device.DeviceID, device.ClientIP, device.UserAgent,
+		device.CreatedAt.UTC(), device.UpdatedAt.UTC()}
+	switch s.sqlStore.GetDBType() {
+	case migrator.Postgres:
+		query = `INSERT INTO anon_device (device_id, client_ip, user_agent, created_at, updated_at)
+VALUES ($1, $2, $3, $4, $5)
+ON CONFLICT (device_id) DO UPDATE SET
+client_ip = $2,
+user_agent = $3,
+updated_at = $5
+RETURNING id`
+	case migrator.MySQL:
+		query = `INSERT INTO anon_device (device_id, client_ip, user_agent, created_at, updated_at)
+VALUES (?, ?, ?, ?, ?)
+ON DUPLICATE KEY UPDATE
+client_ip = VALUES(client_ip),
+user_agent = VALUES(user_agent),
+updated_at = VALUES(updated_at)`
+	case migrator.SQLite:
+		query = `INSERT INTO anon_device (device_id, client_ip, user_agent, created_at, updated_at)
+VALUES (?, ?, ?, ?, ?)
+ON CONFLICT (device_id) DO UPDATE SET
+client_ip = excluded.client_ip,
+user_agent = excluded.user_agent,
+updated_at = excluded.updated_at`
+	default:
+		return fmt.Errorf("unsupported database driver: %s", s.sqlStore.GetDBType())
+	}
+
+	_, err := s.sqlStore.GetSqlxSession().Exec(ctx, query, args...)
+	return err
+}
+
+func (s *AnonDBStore) CountDevices(ctx context.Context, from time.Time, to time.Time) (int64, error) {
+	var count int64
+	err := s.sqlStore.GetSqlxSession().Get(ctx, &count, "SELECT COUNT(*) FROM anon_device WHERE updated_at BETWEEN ? AND ?", from.UTC(), to.UTC())
+	if err != nil {
+		return 0, err
+	}
+	return count, nil
+}
+
+func (s *AnonDBStore) DeleteDevice(ctx context.Context, deviceID string) error {
+	_, err := s.sqlStore.GetSqlxSession().Exec(ctx, "DELETE FROM anon_device WHERE device_id = ?", deviceID)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// deleteOldDevices deletes all devices that have no been updated since the given time.
+func (s *AnonDBStore) deleteOldDevices(ctx context.Context, olderThan time.Time) error {
+	_, err := s.sqlStore.GetSqlxSession().Exec(ctx, "DELETE FROM anon_device WHERE updated_at <= ?", olderThan.UTC())
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *AnonDBStore) Run(ctx context.Context) error {
+	ticker := time.NewTicker(2 * time.Hour)
+
+	for {
+		select {
+		case <-ticker.C:
+			err := s.serverLock.LockAndExecute(ctx, "cleanup old anon devices", time.Hour*10, func(context.Context) {
+				if err := s.deleteOldDevices(ctx, time.Now().Add(-keepFor)); err != nil {
+					s.log.Error("An error occurred while deleting old anon devices", "err", err)
+				}
+			})
+			if err != nil {
+				s.log.Error("Failed to lock and execute cleanup old anon devices", "error", err)
+			}
+
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
```

Comment:
We are running this job every 2 hours but locking the execution of it to 10 hours, won't that cause some error logs when next couple of jobs are scheduled?